src/debugpy/server/tracing/tracer.py

```diff
@@ -0,0 +1,417 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+# Once callbacks are registered they are invoked even during finalization when the
+# Python is shutting down. Thus, trace_* methods, and any other methods that they
+# invoke, must not use any globals from this or other modules (including globals
+# that represent imported modules or defined classes!) until it checks that they
+# are present, or preload them into class or instance attributes in advance.
+# To facilitate this, Tracer is defined in a separate submodule which should not
+# contain ANY top-level imports other than typing nor definitions other than the
+# class itself. All other imports must be done in class scope and then referred to
+# from methods via self.
+
+from types import CodeType, FrameType
+from typing import Iterable
+
+
+class Log:
+    """
+    Safe logging for Tracer. Delegates to debugpy.common.log, but only when it is
+    safe to do so (i.e. not during finalization).
+    """
+
+    from debugpy.common import log
+
+    def __init__(self):
+        import atexit
+
+        def nop(*args, **kwargs):
+            pass
+
+        @atexit.register
+        def disable():
+            self.debug = self.info = self.warning = self.error = self.exception = nop
+
+    def debug(self, *args, **kwargs):
+        # self.log.debug("{0}", *args, **kwargs)
+        # print(*args)
+        pass
```

Comment:
Is this intentional? Any reason to keep the commented out code? If it is used sometimes, should it be conditional on a flag?

Reply:
I'm still trying to figure out a good way to log at this level of detail without a significant perf penalty.

Reply:
Dump to a string buffer with occasional flushes?

Reply:
Yep, I'm going to try that, but as a separate PR.